core/src/main/java/org/lflang/analyses/statespace/Event.java

```diff
@@ -0,0 +1,44 @@
+/** A node in the state space diagram representing a step in the execution of an LF program. */
+package org.lflang.analyses.statespace;
+
+import org.lflang.generator.TriggerInstance;
+
+public class Event implements Comparable<Event> {
+
+  public TriggerInstance trigger;
+  public Tag tag;
+
+  public Event(TriggerInstance trigger, Tag tag) {
+    this.trigger = trigger;
+    this.tag = tag;
+  }
+
+  public TriggerInstance getTrigger() {
+    return this.trigger;
+  }
+
+  @Override
+  public int compareTo(Event e) {
+    // Compare tags first.
+    int ret = this.tag.compareTo(e.tag);
+    // If tags match, compare trigger names.
+    if (ret == 0) ret = this.trigger.getFullName().compareTo(e.trigger.getFullName());
+    return ret;
+  }
+
+  /** This equals() method does NOT compare tags, only compares triggers. */
```

Comment:
You might consider making this a method of its own that does not override `equals` for this reason. See the [`Comparable` contract](https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html). Data structures in the Java standard library (probably including `PriorityQueue`) assume that the ordering induced by `compareTo` is total with respect to the equality defined by `equals`.

Reply:
Thanks for pointing this out. Yes, I will come up with another way to implement this without overriding `equals()` here because it is inconsistent with `compareTo()` above.

By the way, I recently found that it is a general rule that every time `equals()` is overridden, `hashCode()` also needs to be overridden. I found a few instances in the code base where a new `hashCode()` is not provided.